README.md
# MicroPython EEPROM

[![Downloads](https://pepy.tech/badge/micropython-eeprom)](https://pepy.tech/project/micropython-eeprom)
![Release](https://img.shields.io/github/v/release/brainelectronics/micropython-eeprom?include_prereleases&color=success)
![MicroPython](https://img.shields.io/badge/micropython-Ok-green.svg)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![codecov](https://codecov.io/github/brainelectronics/micropython-eeprom/branch/main/graph/badge.svg)](https://app.codecov.io/github/brainelectronics/micropython-eeprom)
[![CI](https://github.com/brainelectronics/micropython-eeprom/actions/workflows/release.yml/badge.svg)](https://github.com/brainelectronics/micropython-eeprom/actions/workflows/release.yml)

MicroPython driver for AT24Cxx EEPROM

---------------

## General

MicroPython driver for AT24Cxx EEPROM

📚 The latest documentation is available at
[MicroPython EEPROM ReadTheDocs][ref-rtd-micropython-eeprom] 📚

<!-- MarkdownTOC -->

- [Installation](#installation)
    - [Install required tools](#install-required-tools)
- [Setup](#setup)
    - [Install package with upip](#install-package-with-upip)
        - [General](#general)
        - [Specific version](#specific-version)
        - [Test version](#test-version)
    - [Manually](#manually)
        - [Upload files to board](#upload-files-to-board)
- [Usage](#usage)
- [Contributing](#contributing)
    - [Unittests](#unittests)
- [Credits](#credits)

<!-- /MarkdownTOC -->

## Installation

### Install required tools

Python3 must be installed on your system. Check the current Python version
with the following command

```bash
python --version
python3 --version
```

Depending on which command `Python 3.x.y` (with x.y as some numbers) is
returned, use that command to proceed.

```bash
python3 -m venv .venv
source .venv/bin/activate

pip install -r requirements.txt
```

## Setup

### Install package with upip

Connect the MicroPython device to a network (if possible)

```python
import network
station = network.WLAN(network.STA_IF)
station.active(True)
station.connect('SSID', 'PASSWORD')
station.isconnected()
```

#### General

Install the latest package version of this lib on the MicroPython device

```python
import mip
mip.install("github:brainelectronics/micropython-eeprom")
```

For MicroPython versions below 1.19.1 use the `upip` package instead of `mip`

```python
import upip
upip.install('micropython-eeprom')
```

#### Specific version

Install a specific, fixed package version of this lib on the MicroPython device

```python
import mip
# install a verions of a specific branch
mip.install("github:brainelectronics/micropython-eeprom", version="feature/initial-implementation")
# install a tag version
mip.install("github:brainelectronics/micropython-eeprom", version="0.1.0")
```

For MicroPython versions below 1.19.1 use the `upip` package instead of `mip`

```python
import upip
upip.install('micropython-eeprom')
```

#### Test version

Install a specific release candidate version uploaded to
[Test Python Package Index](https://test.pypi.org/) on every PR on the
MicroPython device. If no specific version is set, the latest stable version
will be used.

```python
import mip
mip.install("github:brainelectronics/micropython-eeprom", version="0.1.0-rc1.dev1")
```

For MicroPython versions below 1.19.1 use the `upip` package instead of `mip`

```python
import upip
# overwrite index_urls to only take artifacts from test.pypi.org
upip.index_urls = ['https://test.pypi.org/pypi']
upip.install('micropython-eeprom')
```

See also [brainelectronics Test PyPi Server in Docker][ref-brainelectronics-test-pypiserver]
for a test PyPi server running on Docker.

### Manually

#### Upload files to board

Copy the module to the MicroPython board and import them as shown below
using [Remote MicroPython shell][ref-remote-upy-shell]

Open the remote shell with the following command. Additionally use `-b 115200`
in case no CP210x is used but a CH34x.

```bash
rshell --port /dev/tty.SLAB_USBtoUART --editor nano
```

Perform the following command inside the `rshell` to copy all files and
folders to the device

```bash
mkdir /pyboard/lib
mkdir /pyboard/lib/eeprom

cp eeprom/* /pyboard/lib/eeprom

cp examples/main.py /pyboard
cp examples/boot.py /pyboard
```

## Usage

```python
from eeprom import EEPROM
from machine import I2C, Pin

I2C_ADDR = 0x50     # DEC 80, HEX 0x50
EEPROM_SIZE = 32    # AT24C32 on 0x50

# define custom I2C interface, default is 'I2C(0)'
# check the docs of your device for further details and pin infos
i2c = I2C(0, scl=Pin(13), sda=Pin(12), freq=800000)
eeprom = EEPROM(addr=I2C_ADDR, at24x=EEPROM_SIZE, i2c=i2c)

# write 'micropython' to address 10
eeprom.write(10, 'micropython')

# read 11 bytes starting at address 10
eeprom.read(10, 11)

# update content at address 10 with 'MicroPython'
# only changed values are written, here 'm' -> 'M' and 'p' -> 'P'
eeprom.write(10, 'MicroPython')
```

## Contributing

### Unittests

Run the unittests locally with the following command after installing this
package in a virtual environment

```bash
# run all tests
nose2 --config tests/unittest.cfg

# run only one specific tests
nose2 tests.test_eeprom.TestEEPROM.test_addr
```

Generate the coverage files with

```bash
python create_report_dirs.py
coverage html
```

The coverage report is placed at `reports/coverage/html/index.html`

## Credits

Based on
[Mike Causer's MicroPython TinyRTC I2C module][ref-micropython-tinyrtc-i2c]
and the [PyPa sample project][ref-pypa-sample]

<!-- Links -->
[ref-rtd-micropython-eeprom]: https://micropython-eeprom.readthedocs.io/en/latest/
[ref-remote-upy-shell]: https://github.com/dhylands/rshell
[ref-brainelectronics-test-pypiserver]: https://github.com/brainelectronics/test-pypiserver
[ref-micropython-tinyrtc-i2c]: https://github.com/mcauser/micropython-tinyrtc-i2c
[ref-pypa-sample]: https://github.com/pypa/sampleproject
[ref-test-pypi]: https://test.pypi.org/
[ref-pypi]: https://pypi.org/


changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

<!--
## [x.y.z] - yyyy-mm-dd
### Added
### Changed
### Removed
### Fixed
-->
<!--
RegEx for release version from file
r"^\#\# \[\d{1,}[.]\d{1,}[.]\d{1,}\] \- \d{4}\-\d{2}-\d{2}$"
-->

## Released
## [0.1.1] - 2023-06-12
### Fixed
- Usage documentation with more comments and WiFi instructions in root README
- Ignore version of `package.json` file in test workflow

## [0.1.0] - 2023-03-29
### Added
- `eeprom.py` in `eeprom` module
- Examples and documentation

### Changed
- Several updates on setup and config files different than the template repo

### Removed
- Not used files provided with [template repo](https://github.com/brainelectronics/micropython-i2c-lcd)

<!-- Links -->
[Unreleased]: https://github.com/brainelectronics/micropython-eeprom/compare/0.1.1...main

[0.1.1]: https://github.com/brainelectronics/micropython-eeprom/tree/0.1.1
[0.1.0]: https://github.com/brainelectronics/micropython-eeprom/tree/0.1.0


tests/test_eeprom.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-

"""Unittest for MicroPython EEPROM"""

import logging
import sys
import unittest
from io import StringIO
from unittest.mock import Mock, patch

from nose2.tools import params


class Pin(object):
    """Fake MicroPython Pin class"""
    def __init__(self, pin: int, mode: int = -1):
        self._pin = pin
        self._mode = mode
        self._value = 0


class I2C(object):
    """Fake MicroPython I2C class"""
    def __init__(self, id: int, *, scl: Pin, sda: Pin, freq: int = 400000):
        self._id = id
        self._scl = scl
        self._sda = sda
        self._freq = freq

    def writeto(addr: int, buf: bytearray, stop: bool = True) -> int:
        return 1

    def readfrom_mem(addr: int,
                     memaddr: int,
                     nbytes: int,
                     *,
                     addrsize: int = 8) -> int:
        return 1

    def writeto_mem(addr: int,
                    memaddr: int,
                    buf: bytearray,
                    *,
                    addrsize: int = 8):
        pass


# custom imports
sys.modules['machine.I2C'] = I2C
to_be_mocked = [
    'machine',
    'time.sleep_ms', 'time.sleep_us',
]
for module in to_be_mocked:
    sys.modules[module] = Mock()

from eeprom import EEPROM  # noqa: E402


class TestEEPROM(unittest.TestCase):
    """This class describes a TestEEPROM unittest."""

    def setUp(self) -> None:
        """Run before every test method"""
        # define a format
        custom_format = "[%(asctime)s][%(levelname)-8s][%(filename)-20s @" \
                        " %(funcName)-15s:%(lineno)4s] %(message)s"

        # set basic config and level for all loggers
        logging.basicConfig(level=logging.INFO,
                            format=custom_format,
                            stream=sys.stdout)

        # create a logger for this TestSuite
        self.test_logger = logging.getLogger(__name__)

        # set the test logger level
        self.test_logger.setLevel(logging.DEBUG)

        # enable/disable the log output of the device logger for the tests
        # if enabled log data inside this test will be printed
        self.test_logger.disabled = False

        self.i2c = I2C(1, scl=Pin(13), sda=Pin(12), freq=800_000)
        self._calls_counter: int = 0
        self._tracked_call_data: list = []
        self._mocked_read_data: bytes = b''

    def _tracked_call(self, *args, **kwargs) -> None:
        """Track function calls and the used arguments"""
        self._tracked_call_data.append({'args': args, 'kwargs': kwargs})

    def test_init(self) -> None:
        """Test EEPROM init function"""
        eeprom = EEPROM()
        self.assertEqual(eeprom.addr, 0x50)
        self.assertEqual(eeprom.pages, 128)
        self.assertEqual(eeprom.bpp, 32)
        # self.assertEqual(eeprom._i2c._id, 0)

        eeprom = EEPROM(i2c=self.i2c)
        self.assertEqual(eeprom.addr, 0x50)
        self.assertEqual(eeprom.pages, 128)
        self.assertEqual(eeprom.bpp, 32)
        self.assertEqual(eeprom._i2c._id, 1)

        eeprom = EEPROM(at24x=128)
        self.assertEqual(eeprom.addr, 0x50)
        self.assertEqual(eeprom.pages, 256)
        self.assertEqual(eeprom.bpp, 64)
        # self.assertEqual(eeprom._i2c._id, 0)

    @params(
        (0x50),
        (80),
    )
    def test_addr(self, addr: int) -> None:
        """Test address property"""
        eeprom = EEPROM(addr=addr)

        self.assertEqual(eeprom.addr, addr)

    @params(
        (128, 32, 4096),    # pages, bytes per page, total size in bytes
        (512, 64, 32768),
    )
    def test_capacity(self, pages: int, bpp: int, expectation: int) -> None:
        """Test capacity property"""
        eeprom = EEPROM(pages=pages, bpp=bpp)

        self.assertEqual(eeprom.capacity, expectation)

    @params(
        (128),
        (512),
    )
    def test_pages(self, pages: int) -> None:
        """Test pages property"""
        eeprom = EEPROM(pages=pages)

        self.assertEqual(eeprom.pages, pages)

    @params(
        (32),
        (64),
    )
    def test_bpp(self, bpp: int) -> None:
        """Test bpp property"""
        eeprom = EEPROM(bpp=bpp)

        self.assertEqual(eeprom.bpp, bpp)

    @params(
        (128, 32, 4096),    # pages, bytes per page, total size in bytes
        (512, 64, 32768),
    )
    def test_length(self, pages: int, bpp: int, expectation: int) -> None:
        """Test length, alias for capacity, following Arduino implementation"""
        eeprom = EEPROM(pages=pages, bpp=bpp)

        self.assertEqual(eeprom.length(), expectation)

    def test_read(self) -> None:
        """Test reading bytes from EEPROM"""
        eeprom = EEPROM(pages=128, bpp=32, i2c=self.i2c)

        with self.assertRaises(ValueError) as context:
            eeprom.read(addr=eeprom.capacity + 1)

        self.assertEqual(
            str(context.exception),
            "Read address {} outside of device address range {}".
            format(eeprom.capacity + 1, eeprom.capacity)
        )

        # page_addr = eeprom.capacity - 1
        # try:
        #     eeprom.read(addr=page_addr)
        # except ValueError as e:
        #     self.fail(
        #         "Read of last EEPROM address failed unexpectedly: {}".
        #         format(e)
        #     )

        page_addr = eeprom.capacity - 1
        bytes_to_read = 2
        with self.assertRaises(ValueError) as context:
            eeprom.read(addr=page_addr, nbytes=bytes_to_read)

        self.assertEqual(
            str(context.exception),
            "Last read address {} outside of device address range {}".
            format(page_addr + 2, eeprom.capacity)
        )

        page_addr = 10
        with patch.object(I2C, 'readfrom_mem', wraps=self._tracked_call):
            eeprom.read(addr=page_addr)

        self.assertEqual(len(self._tracked_call_data), 1)
        self.assertEqual(self._tracked_call_data[0]['args'],
                         (eeprom.addr, page_addr, 1))

        self._tracked_call_data = []
        bytes_to_read = 5
        with patch.object(I2C, 'readfrom_mem', wraps=self._tracked_call):
            eeprom.read(addr=page_addr, nbytes=bytes_to_read)

        self.assertEqual(len(self._tracked_call_data), 1)
        self.assertEqual(self._tracked_call_data[0]['args'],
                         (eeprom.addr, page_addr, bytes_to_read))

    def test_write(self) -> None:
        """Test writing bytes to EEPROM"""
        eeprom = EEPROM(pages=128, bpp=32, i2c=self.i2c)

        # write to a non existing page
        for addr in [-1, eeprom.capacity + 1]:
            with self.assertRaises(ValueError) as context:
                eeprom.write(addr=addr, buf=bytes([12]))

            self.assertEqual(
                str(context.exception),
                "Write address {} outside of device address range {}".
                format(addr, eeprom.capacity)
            )

        # write more data than fitting on last page
        addr = (eeprom.pages - 1) * eeprom.bpp
        buf = bytes([42] * (eeprom.bpp + 1))
        with self.assertRaises(ValueError) as context:
            eeprom.write(addr=addr, buf=buf)

        self.assertEqual(
            str(context.exception),
            "Last data at {} does not fit into device address range {}".
            format(addr + len(buf), eeprom.capacity)
        )

        # partial page write
        with patch.object(I2C, 'writeto_mem', wraps=self._tracked_call):
            eeprom.write(addr=0, buf=bytes([93]))

        self.assertEqual(len(self._tracked_call_data), 1)
        self.assertEqual(self._tracked_call_data[0]['args'],
                         (eeprom.addr, 0, bytes([93])))

        self._tracked_call_data = []

        # exact full page write
        with patch.object(I2C, 'writeto_mem', wraps=self._tracked_call):
            eeprom.write(addr=0, buf=bytes([93] * eeprom.bpp))

        self.assertEqual(len(self._tracked_call_data), 1)
        self.assertEqual(self._tracked_call_data[0]['args'],
                         (eeprom.addr, 0, bytes([93] * eeprom.bpp)))

        self._tracked_call_data = []

        # overhanging page write
        with patch.object(I2C, 'writeto_mem', wraps=self._tracked_call):
            eeprom.write(addr=0, buf=bytes([93] * (eeprom.bpp + 1)))

        self.assertEqual(len(self._tracked_call_data), 2)
        self.assertEqual(self._tracked_call_data[0]['args'],
                         (eeprom.addr, 0, bytes([93] * eeprom.bpp)))
        self.assertEqual(self._tracked_call_data[1]['args'],
                         (eeprom.addr, eeprom.bpp, bytes([93] * 1)))

        self._tracked_call_data = []

        # with offset (write not to the begin of a page)
        offset = 5
        with patch.object(I2C, 'writeto_mem', wraps=self._tracked_call):
            eeprom.write(addr=offset, buf=bytes([18]))

        self.assertEqual(len(self._tracked_call_data), 1)
        self.assertEqual(self._tracked_call_data[0]['args'],
                         (eeprom.addr, offset, bytes([18])))

        self._tracked_call_data = []

        # with offset + overhanging (write not to the begin of a page)
        offset = 5
        with patch.object(I2C, 'writeto_mem', wraps=self._tracked_call):
            eeprom.write(addr=offset, buf=bytes([18] * (eeprom.bpp - offset)))

        self.assertEqual(len(self._tracked_call_data), 1)
        self.assertEqual(
            self._tracked_call_data[0]['args'],
            (eeprom.addr, offset, bytes([18] * (eeprom.bpp - offset)))
        )

    def _mocked_read_function(self, addr: int, nbytes: int = 1) -> bytes:
        """Mocked EEPROM read function"""
        if isinstance(self._mocked_read_data[addr], bytes):
            if nbytes == 1:
                return self._mocked_read_data[addr]
            else:
                return self._mocked_read_data[addr:addr + nbytes]
        else:
            return self._mocked_read_data[addr: addr + nbytes].encode()

    def test_update_str(self) -> None:
        """Test updating EEPROM cell with string"""
        eeprom = EEPROM(pages=128, bpp=32, i2c=self.i2c)

        # this will be the new data for the EEPROM
        update_buffer = 'hello world'

        # this is the old, existing data in the EEPROM
        self._mocked_read_data = 'hello carl'

        # ensure there is enough data to read
        self._mocked_read_data += \
            '\xff' * (len(update_buffer) - len(self._mocked_read_data))

        # patch the EEPROM read function with the mocked data
        with patch.object(
            EEPROM, 'read', side_effect=self._mocked_read_function
        ):
            with patch.object(I2C, 'writeto_mem', wraps=self._tracked_call):
                eeprom.update(addr=0, buf=update_buffer)

        # difference is 'w', 'o', 'd'
        self.assertEqual(len(self._tracked_call_data), 3)
        for idx, (addr, val) in enumerate(zip([6, 7, 10], [b'w', b'o', b'd'])):
            self.assertEqual(
                self._tracked_call_data[idx]['args'],
                (eeprom.addr, addr, val)
            )

    def test_update_bytes(self) -> None:
        """Test updating EEPROM cell with bytes"""
        eeprom = EEPROM(pages=128, bpp=32, i2c=self.i2c)

        # this will be the new data for the EEPROM
        update_buffer = b'hello world'

        # this is the old, existing data in the EEPROM
        self._mocked_read_data = 'hello carl'

        # ensure there is enough data to read
        self._mocked_read_data += \
            '\xff' * (len(update_buffer) - len(self._mocked_read_data))

        # patch the EEPROM read function with the mocked data
        with patch.object(
            EEPROM, 'read', side_effect=self._mocked_read_function
        ):
            with patch.object(I2C, 'writeto_mem', wraps=self._tracked_call):
                eeprom.update(addr=0, buf=update_buffer)

        # difference is 'w', 'o', 'd'
        self.assertEqual(len(self._tracked_call_data), 3)
        for idx, (addr, val) in enumerate(zip([6, 7, 10], [b'w', b'o', b'd'])):
            self.assertEqual(
                self._tracked_call_data[idx]['args'],
                (eeprom.addr, addr, val)
            )

    def test_update_int(self) -> None:
        """Test updating EEPROM cell with int"""
        eeprom = EEPROM(pages=128, bpp=32, i2c=self.i2c)

        # this will be the new data for the EEPROM
        update_buffer = [1, 93, 42, 23, 255, 7]

        # this is the old, existing data in the EEPROM
        self._mocked_read_data = [
            # 1, 95, 42, 22, 0, 7
            b'\x01', b'_', b'*', b'\x16', b'\x00', b'\x07',
        ]
        # ensure there is enough data to read
        self._mocked_read_data += \
            '\xff' * (len(update_buffer) - len(self._mocked_read_data))

        # patch the EEPROM read function with the mocked data
        with patch.object(
            EEPROM, 'read', side_effect=self._mocked_read_function
        ):
            with patch.object(I2C, 'writeto_mem', wraps=self._tracked_call):
                eeprom.update(addr=0, buf=update_buffer)

        # difference is 93, 23, 255
        print(self._tracked_call_data)
        self.assertEqual(len(self._tracked_call_data), 3)
        for idx, (addr, val) in enumerate(zip([1, 3, 4],
                                              [b']', b'\x17', b'\xff'])):
            self.assertEqual(
                self._tracked_call_data[idx]['args'],
                (eeprom.addr, addr, val)
            )

    def test_wipe(self) -> None:
        """Test wiping EEPROM"""
        eeprom = EEPROM(pages=128, bpp=32, i2c=self.i2c)

        with patch.object(I2C, 'writeto_mem', wraps=self._tracked_call):
            eeprom.wipe()

        self.assertEqual(len(self._tracked_call_data), 128)
        for idx, ele in enumerate(self._tracked_call_data):
            self.assertEqual(
                ele['args'],
                (eeprom.addr, idx * eeprom.bpp, bytes([255] * eeprom.bpp))
            )

    # @unittest.skip('bl')
    def test_print_pages(self) -> None:
        """Test print of EEPROM pages"""
        eeprom = EEPROM(pages=128, bpp=4, i2c=self.i2c)

        # this is the existing data in the EEPROM
        self._mocked_read_data = 'hello carl'

        # patch the EEPROM read function with the mocked data
        with patch.object(
            EEPROM, 'read', side_effect=self._mocked_read_function
        ):
            with patch('sys.stdout', new_callable=StringIO) as mock_stdout:
                eeprom.print_pages(addr=0, nbytes=len(self._mocked_read_data))

        printed_lines = mock_stdout.getvalue().split('\n')
        self.assertEqual(len(printed_lines), 1 + 4)     # first line + x
        self.assertEqual(
            printed_lines[0],
            'Page ---x: 0 --- {}'.format(eeprom.bpp)
        )
        for x in range(0, 3):
            start_pos = x * eeprom.bpp
            end_pos = (x + 1) * eeprom.bpp

            data = self._mocked_read_data[start_pos:end_pos].encode()

            if end_pos > len(self._mocked_read_data):
                data += b'?' * (end_pos - len(self._mocked_read_data))

            self.assertEqual(
                printed_lines[x + 1],
                'Page ---{}: {}'.format(x, data)
            )

    def test_print_pages_offset(self) -> None:
        """Test print of EEPROM pages"""
        eeprom = EEPROM(pages=128, bpp=4, i2c=self.i2c)

        # this is the existing data in the EEPROM
        self._mocked_read_data = 'hello carl'
        offset = 1

        # patch the EEPROM read function with the mocked data
        with patch.object(
            EEPROM, 'read', side_effect=self._mocked_read_function
        ):
            with patch('sys.stdout', new_callable=StringIO) as mock_stdout:
                eeprom.print_pages(
                    addr=offset,
                    nbytes=len(self._mocked_read_data)
                )

        printed_lines = mock_stdout.getvalue().split('\n')
        self.assertEqual(len(printed_lines), 1 + 4)     # first line + x
        self.assertEqual(
            printed_lines[0],
            'Page ---x: 0 --- {}'.format(eeprom.bpp)
        )
        for x in range(0, 3):
            start_pos = x * eeprom.bpp
            end_pos = (x + 1) * eeprom.bpp

            data = self._mocked_read_data[start_pos:end_pos].encode()
            if x == 0:
                data = b'?' * offset + data[offset:]

            if end_pos > len(self._mocked_read_data):
                data += b'?' * (end_pos - len(self._mocked_read_data) - offset)

            self.assertEqual(
                printed_lines[x + 1],
                'Page ---{}: {}'.format(x, data)
            )

    def tearDown(self) -> None:
        """Run after every test method"""
        pass


if __name__ == '__main__':
    unittest.main()


eeprom/eeprom.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-

"""
I2C EEPROM driver for AT24Cxx

EEPROM data sheet: https://ww1.microchip.com/downloads/en/DeviceDoc/doc0336.pdf

MIT License
Copyright (c) 2018 Mike Causer
Extended 2023 by brainelectronics

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""

# system packages
from machine import I2C
from time import sleep_ms


class _Subscriptable():
    def __getitem__(self, item):
        return None


_subscriptable = _Subscriptable()

List = _subscriptable
Optional = _subscriptable
Union = _subscriptable


class EEPROM(object):
    """Driver for AT24Cxx I2C EEPROM"""

    def __init__(self,
                 addr: int = 0x50,
                 pages: int = 128,
                 bpp: int = 32,
                 i2c: Optional[I2C] = None,
                 at24x: int = 0) -> None:
        """
        Constructs a new instance

        :param      addr:   The I2C bus address of the EEPROM
        :type       addr:   int
        :param      pages:  The number of pages of the EEPROM
        :type       pages:  int
        :param      bpp:    The bytes per page
        :type       bpp:    int
        :param      i2c:    I2C object
        :type       i2c:    I2C
        :param      at24x:  The specific AT24Cxx, either 32, 64, 128, 256, 512
        :type       at24x:  int
        """
        self._addr = addr

        standard_eeproms = {
            32: [128, 32],      # 4KiB 32Kbits, 128 pages, 32 bytes/page
            64: [256, 32],      # 8KiB 64Kbits, 256 pages, 32 bytes/page
            128: [256, 64],     # 16KiB 128Kbits, 256 pages, 64 bytes/page
            256: [512, 64],     # 32KiB 256Kbits, 512 pages, 64 bytes/page
            512: [512, 128],    # 64KiB 512Kbits, 512 pages, 128 bytes/page
        }

        if at24x in standard_eeproms:
            self._pages, self._bpp = standard_eeproms[at24x]
        else:
            self._pages = pages
            self._bpp = bpp

        if i2c is None:
            # default assignment, check the docs
            self._i2c = I2C(0)
        else:
            self._i2c = i2c

    @property
    def addr(self) -> int:
        """
        Get the EEPROM I2C bus address

        :returns:   EEPROM I2C bus address
        :rtype:     int
        """
        return self._addr

    @property
    def capacity(self) -> int:
        """
        Get the storage capacity of the EEPROM

        :returns:   EEPROM capacity of the EEPROM in bytes
        :rtype:     int
        """
        return self._pages * self._bpp

    @property
    def pages(self) -> int:
        """
        Get the number of EEPROM pages

        :returns:   Number of pages of the EEPROM
        :rtype:     int
        """
        return self._pages

    @property
    def bpp(self) -> int:
        """
        Get the bytes per page of the EEPROM

        :returns:   Bytes per pages of the EEPROM
        :rtype:     int
        """
        return self._bpp

    def length(self) -> int:
        """
        Get the EEPROM length

        :returns:   Number of cells in the EEPROM
        :rtype:     int
        """
        return self.capacity

    def read(self, addr: int, nbytes: int = 1) -> bytes:
        """
        Read bytes from the EEPROM

        :param      addr:    The start address
        :type       addr:    int
        :param      nbytes:  The number of bytes to read
        :type       nbytes:  int

        :returns:   Data of EEPROM
        :rtype:     bytes
        """
        if addr > self.capacity or addr < 0:
            raise ValueError(
                "Read address {} outside of device address range {}".
                format(addr, self.capacity)
            )

        if addr + nbytes > self.capacity:
            raise ValueError(
                "Last read address {} outside of device address range {}".
                format(addr + nbytes, self.capacity)
            )

        return self._i2c.readfrom_mem(self._addr, addr, nbytes, addrsize=16)

    def write(self, addr: int, buf: Union[bytes, List[int], str]) -> None:
        """
        Write data to the EEPROM

        :param      addr:  The start address
        :type       addr:  int
        :param      buf:   The buffer to write to the EEPROM
        :type       buf:   Union[bytes, List[int], str]
        """
        offset = addr % self._bpp
        partial = 0

        if addr > self.capacity or addr < 0:
            raise ValueError(
                "Write address {} outside of device address range {}".
                format(addr, self.capacity)
            )

        if addr + len(buf) > self.capacity:
            raise ValueError(
                "Last data at {} does not fit into device address range {}".
                format(addr + len(buf), self.capacity)
            )

        # partial page write
        if offset > 0:
            partial = self._bpp - offset
            self._i2c.writeto_mem(
                self._addr, addr, buf[0:partial], addrsize=16
            )
            sleep_ms(5)
            addr += partial

        # full page write
        for i in range(partial, len(buf), self._bpp):
            self._i2c.writeto_mem(
                self._addr,
                addr + i - partial,
                buf[i:i + self._bpp],
                addrsize=16
            )
            sleep_ms(5)

    def update(self, addr: int, buf: Union[bytes, List[int], str]) -> None:
        """
        Update data in EEPROM

        :param      addr:  The start address
        :type       addr:  int
        :param      buf:   The buffer to write to the EEPROM
        :type       buf:   Union[bytes, List[int], str]
        """
        for idx, ele in enumerate(buf):
            this_addr = addr + idx

            if isinstance(ele, int):
                this_val = ele.to_bytes(1, 'big')
            else:
                this_val = str(ele).encode()

            current_value = self.read(addr=this_addr)   # returns bytes
            if current_value != this_val:
                self.write(addr=this_addr, buf=this_val)
            #     print("{}: {} -> {}".
            #           format(this_addr, current_value, this_val))
            # else:
            #     print("No need to update value at {}".format(this_addr))

    def wipe(self) -> None:
        """Wipe the complete EEPROM"""
        page_buff = b'\xff' * self.bpp
        for i in range(self.pages):
            self.write(i * self.bpp, page_buff)

    def print_pages(self, addr: int, nbytes: int) -> None:
        """
        Print pages content with boundaries.

        :param      addr:    The start address
        :type       addr:    int
        :param      nbytes:  The number of bytes to read
        :type       nbytes:  int
        """
        unknown_data_first_page = addr % self.bpp
        unknown_data_last_page = self.bpp - (addr + nbytes) % self.bpp
        if unknown_data_last_page % self.bpp == 0:
            unknown_data_last_page = 0

        data = self.read(addr=addr, nbytes=nbytes)
        extended_data = (
            b'?' * unknown_data_first_page +
            data +
            b'?' * unknown_data_last_page
        )

        sliced_data = [
            extended_data[i: i + self.bpp] for i in range(0,
                                                          len(extended_data),
                                                          self.bpp)
        ]
        print('Page {:->4}: 0 {} {}'.
              format('x', '-' * (self.bpp - len(str(self.bpp))), self.bpp))
        for idx, a_slice in enumerate(sliced_data):
            print('Page {:->4}: {}'.format(idx, a_slice))


examples/main.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-

"""I2C EEPROM showcase"""

from eeprom import EEPROM
from machine import I2C, Pin

# AT24C32 on 0x50
I2C_ADDR = 0x50     # DEC 80, HEX 0x50

# define custom I2C interface, default is 'I2C(0)'
# check the docs of your device for further details and pin infos
# this are the pins for the Raspberry Pi Pico adapter board
i2c = I2C(0, scl=Pin(13), sda=Pin(12), freq=800000)
eeprom = EEPROM(addr=I2C_ADDR, at24x=32, i2c=i2c)

# get LCD infos/properties
print("EEPROM is on I2C address 0x{0:02x}".format(eeprom.addr))
print("EEPROM has {} pages of {} bytes".format(eeprom.pages, eeprom.bpp))
print("EEPROM size is {} bytes ".format(eeprom.capacity))

# write 'micropython' to address 10
print("Save 'micropython' at address 10 ...")
eeprom.write(10, 'micropython')

# read 1 byte from address 4
print("Read 1 byte from address 4")
eeprom.read(4)

# read 10 byte from address 0
print("Read 10 byte from address 0")
eeprom.read(0, 10)

# write multiple pages, starting at address 25
# check bpp property of used EEPROM
# if 32 bytes are used, page 0 will contain "blablubf", page 1 will contain
# the remaining data "oobarbazquxquuxcorgegraultgraplywaldofredplughxyzzythud"
str_long = "blablubfoobarbazquxquuxcorgegraultgraplywaldofredplughxyzzythud"
print("Write '{}' to address 25".format(str_long))
eeprom.write(25, str_long)

# write a list of integers to the EEPROM, starting at address 64
int_list = [5, 17, 255, 9, 12, 7, 255]
print("Write list of integers to address 64: {}".format(int_list))
eeprom.write(64, int_list)

# update only changed values
# here 'm' -> 'M' and 'p' -> 'P', see example above
print("Update EEPROM with 'MicroPython' at address 10")
eeprom.write(10, 'MicroPython')

# update only changed values
# here 5 -> 7, 13 -> 13, 255 -> 244, see example above
int_list_updated = [7, 17, 255, 9, 13, 7, 244]
print("Update EEPROM with '{}' at address 64".format(int_list_updated))
eeprom.update(64, int_list_updated)
